### PR TITLE
meta-iotqa: add qa tests for VAAPI

### PR DESCRIPTION
**Pull request by AshwinSoman**
Add tests for multimedia/vaapi
Signed-off-by: Ashwin Soman <ashwinx.soman@intel.com>

### CODE REVIEW

**Comment by AshwinSoman** [comment]
Please help to test this build.

**Comment by okartau** [comment]
ok to test

File: meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        '''
+    # Decrey H.264 encoded file using VAAPI
+
+    def test_gst_vaapi_can_decode_h264(self):
+        if "vaapih264dec" in self.available_vaapi_elements :
+            (status, output) = self.target.run(
+                'gst-launch-1.0 filesrc location= %s ! qtdemux ! vaapih264dec ! vaapisink' % VAAPITest.h264_encoded_file_target)
+            self.assertEqual(status, 0, msg="Error messages: %s" % output)
+        else :
+            print ("vaapi h264 decode not supported ")
```
Comment (ipuustin) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py]: Or should the test fail instead of printing a message?

**Comment by AshwinSoman** [comment]
Hello All,

NO the test should not fail as every board may not support all the encoders and decoders so if a board doesn’t support a particular encoding or decoding then it doesn’t make sense to fail test case .

Test will fail if board supports any particular encoding and decoding and still it’s not able to do so .

Regards
Ashwin

**Comment by AshwinSoman** [comment]
I have re-based and squashed my commits can you please review and merge it to master . If it's ok to merge please remove test build label too .

**Comment by ipuustin** [comment]
I'll first remove the test-build label if this is ready to be reviewed.

**Comment by AshwinSoman** [comment]
ok

File: meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+'''
+This test suit tests VAAPI Decoding and Encoding
+'''
+
+
+class VAAPITest(oeRuntimeTest):
+
+    '''
+    The video files have been created using gstreamer and videotestsrc as input
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py]: Can they be generated in ```setUpClass()```?

Comment (AshwinSoman) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py]: No as we will need different gstreamer plugins to create then which are not there in image .

**Comment by mythi** [comment]
On Wed, Aug 9, 2017 at 9:14 AM, Ashwin Soman wrote:

> *@AshwinSoman* commented on this pull request.
> ------------------------------
>
> In meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_
> encode_decode.py
> <https://github.com/intel/intel-iot-refkit/pull/279#discussion_r132097651>
> :
>
> > @@ -0,0 +1,198 @@
> +import os
> +from oeqa.oetest import oeRuntimeTest
> +
> +'''
> +This test suit tests VAAPI Decoding and Encoding
> +'''
> +
> +
> +class VAAPITest(oeRuntimeTest):
> +
> +    '''
> +    The video files have been created using gstreamer and videotestsrc as input
>
> No as we will need different gstreamer plugins to create then which are
> not there in image .
>
>
> How is it possible you have the encoding tests then? AFAIK those encoding
tests are creating exactly the same files.

**Comment by AshwinSoman** [comment]
Yes you are right they create same files , but as we don't have plugins I have commented them out right now . Once we have plugins we will need to just remove comment and It will be good to go .

File: meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py (hunk 2 above)
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py]: You can simplify the code quite a bit if you do the following:
- on the target check for available encoders and decoders
- for each format: check both encoder and decoder exists, encode a sample file and then decode it
- if the check is false, skip the test for that platform

File: meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+                os.path.dirname(__file__),
+                'files',
+                VAAPITest.vp8_encoded_file),
+            VAAPITest.vp8_encoded_file_target)
+
+        # get supported vaapi encoders and decoders
+
+        (status, output) = self.target.run(" gst-inspect-1.0 | grep vaapi")
+        output = output.lower()
+        self.available_vaapi_elements = output
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py]: all of the test functions expect ```vaapisink``` which is not checked anywhere. please add a sanity check for it.

File: meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    # Decrey H.264 encoded file using VAAPI
+
+    def test_gst_vaapi_can_decode_h264(self):
+        if "vaapih264dec" in self.available_vaapi_elements :
+            (status, output) = self.target.run(
+                'gst-launch-1.0 filesrc location= %s ! qtdemux ! vaapih264dec ! vaapisink' % VAAPITest.h264_encoded_file_target)
+            self.assertEqual(status, 0, msg="Error messages: %s" % output)
+        else :
+            print ("vaapi h264 decode not supported ")
+    '''
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py]: no commented code please. I've proposed a mechanism how to best detect the platform capabilities.

Comment (AshwinSoman) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_encode_decode.py]: **on the target check for available encoders and decoders**
I am already checking it  

**for each format: check both encoder and decoder exists, encode a sample file and then decode it**
To encode a file we need videotestsrc which is not there , we need to add plugin gstreamer1.0-plugins-base for this , Encoding is not possible until we have this .

**Comment by mythi** [comment]
On Wed, Aug 9, 2017 at 10:41 AM, Ashwin Soman <notifications@github.com>
wrote:

> *@AshwinSoman* commented on this pull request.
> ------------------------------
>
> In meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_
> encode_decode.py
> <https://github.com/intel/intel-iot-refkit/pull/279#discussion_r132111052>
> :
>
> > +            os.path.join(
> +                os.path.dirname(__file__),
> +                'files',
> +                VAAPITest.vp9_encoded_file),
> +            VAAPITest.vp9_encoded_file_target)
> +        '''
> +    # Decrey H.264 encoded file using VAAPI
> +
> +    def test_gst_vaapi_can_decode_h264(self):
> +        if "vaapih264dec" in self.available_vaapi_elements :
> +            (status, output) = self.target.run(
> +                'gst-launch-1.0 filesrc location= %s ! qtdemux ! vaapih264dec ! vaapisink' % VAAPITest.h264_encoded_file_target)
> +            self.assertEqual(status, 0, msg="Error messages: %s" % output)
> +        else :
> +            print ("vaapi h264 decode not supported ")
> +    '''
>
> *on the target check for available encoders and decoders*
> I am already checking it
>

I was just explaining the logic.


> *for each format: check both encoder and decoder exists, encode a sample
> file and then decode it*
> To encode a file we need videotestsrc which is not there , we need to add
> plugin gstreamer1.0-plugins-base for this , Encoding is not possible until
> we have this .
>
>
>
You can add the missing packages in ```packagegroup-computervision-test.bb
```.

**Comment by AshwinSoman** [comment]
please help to review and merge it to master

File: meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_present.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+'''
+This test suit tests VAAPI is present or not
+'''
+
+from oeqa.oetest import oeRuntimeTest
+
+
+class VAAPITest(oeRuntimeTest):
+    def test_vaapi_present(self):
+        (status, output) = self.target.run(" gst-inspect-1.0 | grep vaapi")
```
Comment (mythi) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_present.py]: to simplify the code you could just ```status, output = self.target.run("gst-inspect-1.0 --exists vaapisink")``` and ```self.assertEqual(status, 0...)```

**Comment by AshwinSoman** [comment]
please help to review and merge it to master

**Comment by AshwinSoman** [comment]
@mythi  I have  changed self.target.run(" gst-inspect-1.0 | grep vaapi") to self.target.run("vainfo")
as that was not the correct way to check VAAPI present or not .
Below lines are important because if VAAPI is not there vainfo wont be there so the return status will be 0 as there will be no error . where as test should fail if VAAPI is not there . 
if  "VA-API" in output :
             status = 0
         else :
             status = 1

**Comment by mythi** [comment]
On Tue, Aug 15, 2017 at 10:03 AM, Ashwin Soman <notifications@github.com>
wrote:

> @mythi <https://github.com/mythi> I have changed self.target.run("
> gst-inspect-1.0 | grep vaapi") to self.target.run("vainfo")
> as that was not the correct way to check VAAPI present or not .
>

OK


> Below lines are important because if VAAPI is not there vainfo wont be
> there so the return status will be 0 as there will be no error . where as
> test should fail if VAAPI is not there .
> if "VA-API" in output :
> status = 0
> else :
> status = 1
>

Are you sure about this? On my PC if I remove the driver and run vainfo, I
get non-zero exit status.

File: meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_present.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+'''
+This test suit tests VAAPI is present or not
+'''
+
+from oeqa.oetest import oeRuntimeTest
+
+class VAAPITest(oeRuntimeTest):
+    def test_vaapi_present(self):
+        (status, output) = self.target.run(" vainfo")
```
Comment (ipuustin) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_present.py]: Why space in `" vainfo"`?

Comment (AshwinSoman) [inline meta-iotqa/lib/oeqa/runtime/multimedia/vaapi/test_vaapi_present.py]: space removed

**Pull request merged**